core/node/state_keeper/src/seal_criteria/io_criteria.rs

```diff
@@ -0,0 +1,333 @@
+            "Non-empty L2 block with too recent timestamp shouldn't be sealed"
+        );
+    }
+
+    #[test]
+    fn max_size_l2_block_sealer() {
+        let tx = create_transaction(10, 100);
+        let tx_encoding_size =
+            zksync_protobuf::repr::encode::<zksync_dal::consensus::proto::Transaction>(&tx).len();
+
+        let mut max_payload_sealer = L2BlockMaxPayloadSizeSealer {
+            max_payload_size: tx_encoding_size,
+        };
+
+        let mut manager = create_updates_manager();
+        assert!(
+            !max_payload_sealer.should_seal_l2_block(&manager),
+            "Empty L2 block shouldn't be sealed"
+        );
+
+        apply_tx_to_manager(tx, &mut manager);
+        assert!(
+            max_payload_sealer.should_seal_l2_block(&manager),
+            "L2 block with payload encoding size equal or greater than max payload size should be sealed"
+        );
+    }
+
+    #[tokio::test]
+    async fn protocol_upgrade_sealer() {
+        const CHECK_INTERVAL: Duration = Duration::from_millis(100);
+
+        let pool = ConnectionPool::<Core>::test_pool().await;
+        let mut sealer = ProtocolUpgradeSealer::new().with_check_interval(CHECK_INTERVAL);
+
+        let mut conn = pool.connection().await.unwrap();
+
+        conn.protocol_versions_dal()
+            .save_protocol_version_with_tx(&ProtocolVersion {
+                version: ProtocolSemanticVersion {
+                    minor: ProtocolVersionId::latest(),
+                    patch: 0.into(),
+                },
+                ..Default::default()
+            })
+            .await
+            .unwrap();
+
+        let manager = create_updates_manager();
+
+        // No new version, shouldn't seal.
+        let should_seal = sealer
+            .should_seal_l1_batch_unconditionally(&manager, &pool, ProtocolVersionId::latest())
+            .await
+            .unwrap();
+        assert!(!should_seal);
+
+        // Insert new protocol version.
+        conn.protocol_versions_dal()
+            .save_protocol_version_with_tx(&ProtocolVersion {
+                version: ProtocolSemanticVersion {
+                    minor: ProtocolVersionId::next(),
+                    patch: 0.into(),
+                },
+                ..Default::default()
+            })
+            .await
+            .unwrap();
+
+        let first_checked_at = sealer.last_checked_at.unwrap();
+        let should_seal = sealer
+            .should_seal_l1_batch_unconditionally(&manager, &pool, ProtocolVersionId::latest())
+            .await
+            .unwrap();
+        if first_checked_at == sealer.last_checked_at.unwrap() {
+            // This is the expected path, execution should take <CHECK_INTERVAL and sealer should return false.
+            assert!(!should_seal);
+
+            // Reset `last_checked_at` and re-check.
+            sealer.last_checked_at = Some(first_checked_at - 2 * CHECK_INTERVAL);
```

Comment:
Bikeshedding: I'd set this from the start to have a controlled flow of the test (i.e., avoid the enclosing `if`).